### PR TITLE
PKP Mind Virus: Glorfing Cigs edition

### PR DESCRIPTION
**Pull request by LemonInTheDark**
## About The Pull Request

PKP made a forum thread about choking on cigarettes if you glorfed while smoking. I thought it was really funny so I've spent the last 3 days working on choke code.

If you glorf while smoking you'll get the cig stuck in your throat, leaving you unable to breath, speak, or eat.
Based off the sort of cig you have a random delay before you vomit up the cig. (Pipes last forever don't philoso-larp kids) Other players can help you by using the help action on you while you're hurt, which leads to something imitating the himelich manuver, a broken chest, and a vomited up blockage.

Other forms of vomiting will work too, if that's even applicable.

Lemme see your screen shakes while doin this, and if you swallow a lit cig you spit ash...

Oh I fixed a bug in vomit code. Because paralyzing you forces you to the ground, thus changing your dir, paralyzing vomits always went down. This is cringe, easy fix tho.

Ah and I added some extra functionality to sound loops too. You can set them to pick and take from the provided list, then refilling it once it's empty. This keeps some variety while preventing the risk of repeat sounds

Oh and I added some new signals. Some misc ones, for force say and a different kind of help then the existing one, which I renamed. One for breathing, which I used to clean up leroxin. I was planning to hook this but didn't need to in the end.

One for vomiting too. and eating. Also added a way to update looping timer delays. Forgot about that, I also added variable delay for sound loops, because I thought it sounded better here.

Features audio courtesy of our very own san7890, and the quake jump sound.

## Why It's Good For The Game

https://user-images.githubusercontent.com/58055496/190132710-a93387b6-5e1a-491f-acbb-1124d913ce6c.mp4
The weak should fear the strong.

Now to fix that critical atmos bug

NOTE:
Bumps byond and spacemandmm to the latest version, so they will actually accept the particle param I'm using here.
The effect was added in 514.1545. Attempting to join the server in a version older then that will just fail, so it should be fineeee

## Changelog

:cl:
add: You can now choke on cigarettes. Careful, those suckers can be nasty
/:cl:

### CODE REVIEW

**Comment by Time-Green** [comment]
What's glorfing?

**Comment by Time-Green** [comment]
Glorfing is only input box right, not popup box?

File: code/game/machinery/computer/crew.dm [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /// How often the sensor data is updated
-#define SENSORS_UPDATE_PERIOD 10 SECONDS //How often the sensor data updates.
+#define SENSORS_UPDATE_PERIOD 10 SECONDS //How often the sensor data updates./datum/component/usb_port
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/game/machinery/computer/crew.dm:2]: I uh, I thought I removed that

**Comment by Wallemations** [comment]
What the fuck is glorfing

File: dependencies.sh [hunk 2]
```diff
@@ -5,7 +5,7 @@
 
 # byond version
 export BYOND_MAJOR=514
-export BYOND_MINOR=1560
+export BYOND_MINOR=1588
```
Comment (san7890) [inline dependencies.sh:8]: What?

Comment (LemonInTheDark) [reply dependencies.sh:8]: oh right sorry

**Comment by necromanceranne** [comment]
> What the fuck is glorfing

Something so unlikely to happen its dubious if it works. You get smacked with a stun the _moment_ you send a message.

**Comment by Jacquerel** [comment]
> > What the fuck is glorfing
> 
> Something so unlikely to happen its dubious if it works. You get smacked with a stun the _moment_ you send a message.

You can get glorfed if you have any text at all in the text window and it will dump the contents, it's just that people like to talk in two word sentences rather than writing an essay in there. 

If you can see the typing indicator then whacking that person can glorf them.

**Comment by Wallemations** [comment]
> > > What the fuck is glorfing
> 
> > 
> 
> > Something so unlikely to happen its dubious if it works. You get smacked with a stun the _moment_ you send a message.
> 
> 
> 
> You can get glorfed if you have any text at all in the text window and it will dump the contents, it's just that people like to talk in two word sentences rather than writing an essay in there. 
> 
> 
> 
> If you can see the typing indicator then whacking that person can glorf them. 

Oh it's a normal thing then

It sounded like a weird emote reference to something

**Comment by JohnFulpWillard** [comment]
draft for reviews

**Comment by LemonInTheDark** [comment]
I need to figure out what happens when you run this with an outdated version

**Comment by LemonInTheDark** [comment]
I don't think you can even join the game on this version. Sooooooo it should be fine

File: code/datums/mood_events/generic_negative_events.dm [hunk 3]
```diff
@@ -154,6 +154,10 @@
 	description = "I'm nervous and on edge and I can't stand still!!"
 	mood_change = -2
 
+/datum/mood_event/choke
+	description = "I CAN'T BREATH!!!"
```
Comment (san7890) [inline code/datums/mood_events/generic_negative_events.dm:158]:
```suggestion
	description = "I CAN'T BREATHE!!!"
```

although "i can't breath" is quite funny so feel free to ignore this

Comment (LemonInTheDark) [reply code/datums/mood_events/generic_negative_events.dm:158]: fcuk

File: code/datums/status_effects/debuffs/choke.dm [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+
+	if(!thrusting_continues(victim, aggressor, before_work = TRUE))
+		return
+	aggressor.start_pulling(victim)
+	// I want the thing to look vaugely like the heimlich. Sue me
+	victim.setDir(aggressor.dir)
+
+	var/hand_name = "hand"
+	if(!iscarbon(aggressor))
+		hand_name = "paw" // Fuck you
```
Comment (JohnFulpWillard) [inline code/datums/status_effects/debuffs/choke.dm:196]: who would even call this? gorillas?

File: code/datums/status_effects/debuffs/choke.dm [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+			span_boldwarning("You feel [aggressor]\s [hand_name]s on your back, and then repeated striking!"))
+
+	if(!do_after_mob(aggressor, victim, 7 SECONDS, extra_checks = CALLBACK(src, .proc/thrusting_continues, victim, aggressor), interaction_key = "heimlich"))
+		aggressor.stop_pulling()
+		return
+	aggressor.stop_pulling()
+
+	var/atom/movable/choking_on = choking_on_ref?.resolve()
+	owner.visible_message(span_green("[victim] vomits up \the[choking_on]. [victim.p_theyre()] gonna make it!"), \
+			span_green("You vomit up that accursed blockage. YOU CAN BREATH! The broken chest is a hell of a price to pay."))
```
Comment (JohnFulpWillard) [inline code/datums/status_effects/debuffs/choke.dm:214]:
```suggestion
			span_green("You vomit up that accursed blockage. YOU CAN BREATHE! The broken chest is a hell of a price to pay."))
```

File: code/game/objects/items/cigs_lighters.dm [hunk 6]
```diff
@@ -180,6 +183,21 @@ CIGARETTE PACKETS ARE IN FANCY.DM
 	STOP_PROCESSING(SSobj, src)
 	return ..()
 
+/obj/item/clothing/mask/cigarette/equipped(mob/equipee, slot)
+	. = ..()
+	if(slot != ITEM_SLOT_MASK)
```
Comment (JohnFulpWillard) [inline code/game/objects/items/cigs_lighters.dm:188]:
```suggestion
	if(!(slot & ITEM_SLOT_MASK))
```
i had changed all of these please don't do this to me

File: code/datums/status_effects/debuffs/choke.dm [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+	vomit_up()
+
+/datum/status_effect/choke/proc/mirror_dir(atom/source, old_dir, new_dir)
+	SIGNAL_HANDLER
+	owner.dir = new_dir
+
+/datum/status_effect/choke/proc/thrusting_continues(mob/living/victim, mob/aggressor, before_work = FALSE)
+	if(iscarbon(aggressor))
+		var/free_hands = 0
+		// Listen bud, you need at least 2 free hands for this
```
Comment (JohnFulpWillard) [inline code/datums/status_effects/debuffs/choke.dm:230]: i became cpr certified a month ago and now i realize i hate you

File: code/datums/status_effects/debuffs/choke.dm [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+	return COMPONENT_BLOCK_MISC_HELP
+
+/datum/status_effect/choke/proc/heimlich(mob/victim, mob/aggressor)
+	if(victim == aggressor)
+		return
+	if(DOING_INTERACTION_WITH_TARGET(aggressor, victim))
+		victim.balloon_alert(aggressor, "You are already helping!")
+		return
+	if(DOING_INTERACTION(aggressor, "heimlich"))
+		victim.balloon_alert(aggressor, "You can't heimlich two people at once!") // Yes I know it's kinda runon, but this is a real rare condition
```
Comment (JohnFulpWillard) [inline code/datums/status_effects/debuffs/choke.dm:185]:
```suggestion
		victim.balloon_alert(aggressor, "already helping someone!") // Yes I know it's kinda runon, but this is a real rare condition
```

File: code/datums/status_effects/debuffs/choke.dm [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+/datum/status_effect/choke/proc/shook(mob/source, mob/helping)
+	SIGNAL_HANDLER
+	INVOKE_ASYNC(src, .proc/heimlich, source, helping)
+	return COMPONENT_BLOCK_MISC_HELP
+
+/datum/status_effect/choke/proc/heimlich(mob/victim, mob/aggressor)
+	if(victim == aggressor)
+		return
+	if(DOING_INTERACTION_WITH_TARGET(aggressor, victim))
+		victim.balloon_alert(aggressor, "You are already helping!")
```
Comment (JohnFulpWillard) [inline code/datums/status_effects/debuffs/choke.dm:182]:
```suggestion
		victim.balloon_alert(aggressor, "already helping!")
```

File: code/datums/status_effects/debuffs/choke.dm [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+	check_audio_state()
+
+/datum/status_effect/choke/proc/on_revive(mob/living/source)
+	SIGNAL_HANDLER
+	UnregisterSignal(source, COMSIG_LIVING_REVIVE)
+	check_audio_state()
+
+/datum/status_effect/choke/proc/attempt_eat(mob/source, atom/eating)
+	SIGNAL_HANDLER
+	source.balloon_alert(source, "you can't get it down!")
```
Comment (JohnFulpWillard) [inline code/datums/status_effects/debuffs/choke.dm:165]:
```suggestion
	source.balloon_alert(source, "can't get it down!")
```

File: code/datums/status_effects/debuffs/choke.dm [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+				continue
+			free_hands += 1
+		if(free_hands < 2)
+			victim.balloon_alert(aggressor, "need 2 free hands!")
+			return FALSE
+
+	if(iscarbon(victim))
+		var/mob/living/carbon/carbon_victim = victim
+		if(!carbon_victim.appears_alive())
+			victim.balloon_alert(aggressor, "cannot be dead!")
```
Comment (JohnFulpWillard) [inline code/datums/status_effects/debuffs/choke.dm:242]:
```suggestion
			victim.balloon_alert(aggressor, "too late...")
```
this is funny it should make the user forcesay it or something

File: code/datums/mood_events/generic_negative_events.dm (hunk 3 above)
Comment (JohnFulpWillard) [reply code/datums/mood_events/generic_negative_events.dm:158]: hii

File: code/datums/status_effects/debuffs/choke.dm (hunk 8 above)
Comment (san7890) [reply code/datums/status_effects/debuffs/choke.dm:185]: may as well remove the comment at that rate as well if "run-on" refers to the the text the balloon_alert shows

File: code/datums/status_effects/debuffs/choke.dm (hunk 4 above)
Comment (LemonInTheDark) [reply code/datums/status_effects/debuffs/choke.dm:196]: I think like aliens? dogs maybe?

File: code/datums/status_effects/debuffs/choke.dm (hunk 7 above)
Comment (LemonInTheDark) [reply code/datums/status_effects/debuffs/choke.dm:230]: :)

**Comment by Mothblocks** [comment]
@JohnFulpWillard Please act on this PR if you are interested in merging it

**Pull request merged**